toucan_connectors/google_big_query/google_big_query_connector.py

```diff
@@ -226,21 +232,109 @@ def _format_columns(x: str):
         return '\nUNION ALL\n'.join(
             self._build_dataset_info_query_for_ds(dataset_id, db_name) for dataset_id in dataset_ids
         )
 
+    def _fetch_query_results_per_start_index(
+        self,
+        query: str,
+        client: bigquery.Client,
+        start_index: int,
+        query_job: QueryJob,
+        results: list[Any],
+        lock: threading.Lock,
+    ) -> None:
+        """Fetch query results for a specific start index and append to the shared results list."""
+        query_config = bigquery.QueryJobConfig()
+        query_config.query_parameters = [
+            bigquery.ScalarQueryParameter('startIndex', 'INT64', start_index)
+        ]
+
+        # Create a new query job with the updated configuration
+        query_job = client.query(query, job_config=query_config)
+
+        df_list = [df for df in self._fetch_query_results(query_job)]
+
+        # Lock the results list before appending to avoid race conditions
+        lock.acquire()
+        try:
+            results.append(df_list)
+        finally:
+            lock.release()
+
+    def _merge_query_results(self, query: str, client: bigquery.Client, num_threads: int = 6):
+        """Execute query with multiple threads, incrementing startIndex, and merge results."""
+        lock = threading.Lock()
+
+        results, threads = [], []
+
+        # Create the query job with the initial startIndex
+        # Get the total rows count
+        query_job = client.query(query)
+        total_rows = getattr(query_job, 'total_rows', _MAXIMUM_RESULTS_FETCHED)
+
+        # Calculate the offset based on the total rows count and the number of threads
+        offset = total_rows // num_threads
+
+        # Create and start a thread for each startIndex offset
+        for i in range(num_threads):
+            thread = threading.Thread(
+                target=self._fetch_query_results_per_start_index,
+                args=(query, client, i * offset, query_job, results, lock),
+            )
+            threads.append(thread)
+
+        for thread in threads:
+            thread.start()
+
+        # Wait for all threads to finish
+        for thread in threads:
+            thread.join()
+
+        return pd.DataFrame(results)
+
+    def _fetch_query_results(self, query_job: QueryJob) -> Generator[Any, Any, Any]:
+        """Fetches query results in a paginated manner using a generator."""
+
+        try:
+            row_iterator = query_job.result(
+                page_size=_PAGE_SIZE,
+                max_results=_MAXIMUM_RESULTS_FETCHED,
+            )
+
+            while rows := next(row_iterator.pages, None):
+                yield rows.to_dataframe()
+        except GoogleExceptionNotfound as excp:
+            # we ignore when we don't found a dataset on a locationk
+            # google.api_core.exceptions.NotFound:
+            # 404 Not found: Dataset data-finance-staging:dbt_yanni_iyeze_dbt_test__audit was not found in location EU
```

Comment:
Can you please anonymize the dataset name ?

Also, I beleive that the exception must not be caught here: For clients with datasets in several regions, only those from the first encountered region will appear (and `_get_project_structure_slow` will never be called)